crates/lace-bin/src/typechecking/context.rs

```diff
@@ -0,0 +1,174 @@
+    pub variable_generator: TypeVariableGenerator,
+
+    /// stores constraints. Constraints can only reference type variables,
+    /// not concrete types. This makes the union find phase quicker.
+    pub constraints: Vec<(Constraint, ConstraintMetadata<'a>)>,
+
+    /// Stores a mapping from identifiers to type variables
+    pub name_mapping: NameMapping,
+    /// Stores a mapping from type variables to concrete types
+    pub type_mapping: TypeMapping<'a>,
+}
+
+impl<'a> TypeContext<'a> {
+    pub fn new(arena: &'a Bump) -> Self {
+        Self {
+            variable_generator: TypeVariableGenerator::new(),
+            arena,
+            constraints: vec![],
+            name_mapping: Default::default(),
+            type_mapping: Default::default(),
+        }
+    }
+
+    fn get_or_insert_name_mapping(&mut self, node_id: NodeId) -> TypeVariable {
+        match self.name_mapping.entry(node_id) {
+            Entry::Occupied(o) => *o.get(),
+            Entry::Vacant(v) => {
+                let variable = self.variable_generator.next();
+                *v.insert(variable)
+            }
+        }
+    }
+
+    pub fn add_name_resolutions(&mut self, name_resolutions: &NameResolutions) {
+        for (a, b) in &name_resolutions.names {
+            let a = self.get_or_insert_name_mapping(*a);
+            let b = self.get_or_insert_name_mapping(*b);
+            self.add_equal_constraint(a, b, ConstraintMetadata::NameRef);
+        }
+    }
+
+    pub fn fresh(&mut self) -> TypeVariable {
+        self.variable_generator.next()
+    }
+
+    pub fn concrete_type(&mut self, ty: ConcreteType<'a>) -> TypeVariable {
+        let var = self.fresh();
+        self.type_mapping.insert(var, ty);
+        var
+    }
+
+    pub fn alloc<T>(&self, value: T) -> &'a T {
+        self.arena.alloc(value)
+    }
+
+    pub fn add_equal_constraint_concrete(
+        &mut self,
+        a: TypeVariable,
+        b: ConcreteType<'a>,
+        meta: ConstraintMetadata<'a>,
+    ) {
+        let b = self.concrete_type(b);
+        self.add_equal_constraint(a, b, meta);
+    }
+
+    pub fn add_equal_constraint(
+        &mut self,
+        a: TypeVariable,
+        b: TypeVariable,
+        meta: ConstraintMetadata<'a>,
+    ) {
+        self.constraints.push((Constraint::Equal(a, b), meta));
+    }
+
+    pub fn type_of_name(&mut self, ident: &Identified<Ident>) -> TypeVariable {
+        self.get_or_insert_name_mapping(ident.node_id)
+    }
+
+    fn name_of_type_var(
+        &self,
+        inverse_name_mapping: &HashMap<&TypeVariable, &NodeId>,
+        spans: &Spans,
+        source: SourceFile,
+        var: TypeVariable,
+    ) -> String {
+        if let Some(i) = inverse_name_mapping.get(&var) {
+            let span = spans.get(**i);
+            format!("variable {}", source.slice_span(span))
+        } else if let Some(i) = self.type_mapping.get(&var) {
+            match i {
+                ConcreteType::Int => "int".to_string(),
```

Comment:
You have this implemented elsewhere too

Reply:
Different purposes actually, but the code is rather similar

Reply:
the code is way more recursive in it's lookup, and I want it that way here and not there